### ISSUE TITLE
Square brackets in config don't work anymore in 6.x

### ISSUE THREAD

**Issue by gmoskovicz** [elastic/logstash#9036]
- Version: Logstash 6.1
- Operating System: Any
- Config File (if you have sensitive info, please remove it):
```
input {
	stdin {}
}


filter {
	if [anything] == [] {
	}
}



output {
	stdout { codec => rubydebug }
}

```
- Sample Data: Any data
- Steps to Reproduce:

Just run this and you get:

> [2018-01-25T16:52:42,133][ERROR][logstash.agent           ] Failed to execute action {:action=>LogStash::PipelineAction::Create/pipeline_id:main, :exception=>"Java::JavaLang::IllegalArgumentException", :message=>"Failed to parse right-hand side of conditional [str]pipeline:12:5:```\n[clicksPermitted] == []\n```", :backtrace=>["org.logstash.config.ir.expression.BinaryBooleanExpression.ensureNotNull(org/logstash/config/ir/expression/BinaryBooleanExpression.java:57)", "org.logstash.config.ir.expression.BinaryBooleanExpression.<init>(org/logstash/config/ir/expression/BinaryBooleanExpression.java:37)", "org.logstash.config.ir.expression.binary.Eq.<init>(org/logstash/config/ir/expression/binary/Eq.java:12)", "org.logstash.config.ir.DSL.eEq(org/logstash/config/ir/DSL.java:113)", "java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:497)", "org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:423)", "org.jruby.javasupport.JavaMethod.invokeStaticDirect(org/jruby/javasupport/JavaMethod.java:355)", "org.jruby.RubyMethod.call(org/jruby/RubyMethod.java:127)", "org.jruby.RubyMethod$INVOKER$i$call.call(org/jruby/RubyMethod$INVOKER$i$call.gen)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.expr(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb:438)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.RUBY$method$expr$0$__VARARGS__(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.expr(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb:413)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.RUBY$method$expr$0$__VARARGS__(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.block in join_conditions(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb:360)", "org.jruby.RubyArray.each(org/jruby/RubyArray.java:1734)", "org.jruby.RubyArray$INVOKER$i$0$0$each.call(org/jruby/RubyArray$INVOKER$i$0$0$each.gen)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.join_conditions(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb:359)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.RUBY$method$join_conditions$0$__VARARGS__(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.expr(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb:308)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.RUBY$method$expr$0$__VARARGS__(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.expr_cond(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb:277)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.RUBY$method$expr_cond$0$__VARARGS__(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.block in expr(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb:226)", "org.jruby.RubyArray.each(org/jruby/RubyArray.java:1734)", "org.jruby.RubyArray$INVOKER$i$0$0$each.call(org/jruby/RubyArray$INVOKER$i$0$0$each.gen)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.expr(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb:223)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.RUBY$method$expr$0$__VARARGS__(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb)", "org.jruby.RubySymbol$SymbolProcBody.yieldInner(org/jruby/RubySymbol.java:1041)", "org.jruby.RubySymbol$SymbolProcBody.doYield(org/jruby/RubySymbol.java:1057)", "org.jruby.RubyArray.collect(org/jruby/RubyArray.java:2472)", "org.jruby.RubyArray.map(org/jruby/RubyArray.java:2486)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.expr(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb:69)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.RUBY$method$expr$0$__VARARGS__(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.block in compile(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb:48)", "org.jruby.RubyArray.each(org/jruby/RubyArray.java:1734)", "org.jruby.RubyArray$INVOKER$i$0$0$each.call(org/jruby/RubyArray$INVOKER$i$0$0$each.gen)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.lscl.compile(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash/compiler//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler/lscl.rb:46)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.compile_imperative(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler.rb:46)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.RUBY$method$compile_imperative$0$__VARARGS__(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler.rb)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.compile_graph(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler.rb:50)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.RUBY$method$compile_graph$0$__VARARGS__(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler.rb)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.block in compile_sources(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler.rb:12)", "org.jruby.RubyArray.collect(org/jruby/RubyArray.java:2472)", "org.jruby.RubyArray.map(org/jruby/RubyArray.java:2486)", "org.jruby.RubyArray$INVOKER$i$0$0$map19.call(org/jruby/RubyArray$INVOKER$i$0$0$map19.gen)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.compile_sources(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler.rb:11)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.compiler.RUBY$method$compile_sources$0$__VARARGS__(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/compiler.rb)", "org.jruby.RubyClass.finvoke(org/jruby/RubyClass.java:899)", "org.jruby.RubyBasicObject.callMethod(org/jruby/RubyBasicObject.java:372)", "org.logstash.config.ir.ConfigCompiler.configToPipelineIR(org/logstash/config/ir/ConfigCompiler.java:37)", "java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:497)", "org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:423)", "org.jruby.javasupport.JavaMethod.invokeStaticDirect(org/jruby/javasupport/JavaMethod.java:355)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.pipeline.initialize(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/pipeline.rb:51)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.pipeline.initialize(Users/Gabriel/Documents/ElasticSearch/logstash_minus_6_dot_1_dot_0/logstash_minus_core/lib/logstash//Users/Gabriel/Documents/ElasticSearch/logstash-6.1.0/logstash-core/lib/logstash/pipeline.rb:171)", "org.jruby.RubyClass.newInstance(org/jruby/RubyClass.java:1022)", "org.jruby.RubyClass$INVOKER$i$newInstance.call(org/jruby/RubyClass$INVOKER$i$newInstance.gen)", "Users.Gabriel.Documents.ElasticSearch.logstash_minus_6_dot_1_dot_0.logstash_minus_core.lib.logstash.pi...

**Comment by yaauie** [comment]
It looks like we're hitting a similar bug as https://github.com/elastic/logstash/pull/8755 (in which the LIR can fail to produce an expression on the RHS if the LHS is a regexp), which causes us to hit the "better" error message that we pushed out with https://github.com/elastic/logstash/pull/8774.

Either way, configs with the above syntax will fail to produce a pipeline in 6.0+.

I'll attempt to chase this down to see if we can get a fix in before 6.2.0 is released.

**Comment by gmoskovicz** [comment]
Thanks for the quick review.

**Comment by gmoskovicz** [comment]
@yaauie is there a workaround for now? such as using `[field][0]` for example? I can't think of one :)

**Comment by yaauie** [comment]
@gmoskovicz If you're just looking to execute the block if the field is not set, perhaps this will work?

~~~
input { stdin {} }

filter {
  if ![anything] {
   # ...
  }
}

output { stdout { codec => rubydebug } }
~~~

If you're looking to see if the field contains an empty array though, I don't have an immediate workaround.

**Comment by gmoskovicz** [comment]
Yeah, that doesn't work. It's an empty array but the field does exists. Will wait for the fix :)

**Comment by original-brownbear** [comment]
@yaauie there's gonna be a second problem past fixing the `Graph` class here for the Java execution. It doesn't have any facilities to do array comparison.

@andrewvc @jordansissel comparing arrays isn't documented as far as I can seem do we want to keep supporting this in the Java execution? If so we need to add the logic for that, it's trivial to do that though :)

**Comment by gmoskovicz** [comment]
It seems to be a breaking change in 6.x, although a fix could be to add a conditional logic to check list `length` ?

**Comment by jordansissel** [comment]
Yes, I expect `[] == []` to be true. Similarly, `[foo] == []` should be true iff the `foo` field is an empty array.

**Comment by jordansissel** [comment]
Let's add tests for this and fix the bug.

**Comment by yaauie** [comment]
I have a WIP fix in-flight for supporting _empty_ arrays as comparable values; as @original-brownbear called out, getting the LIR to build correctly is only the surface of the problem, and re-implementing array comparison in the Java pipeline adds a fair bit to the scope, especially in light of some surprising syntax irregularities.

Below are some notes about my deep-dive into `lscl.rb`, the treetop grammar, and the pipeline compiler; all links below use `6.1`, but are also applicable to `master` as of this writing.

tl;dr: there is a lot of mismatch between what expressions will _parse_, which will generate valid LIR, and gaps from there to the actual compiled pipeline.

First, the treetop grammar defines a `ComparisonExpression` as two `rvalue` joined by a `compare_operator`:

> ~~~ ruby
>   rule compare_expression
>     rvalue _ compare_operator _ rvalue
>     <LogStash::Compiler::LSCL::AST::ComparisonExpression>
>   end
> ~~~
> -- [`logstash-core/lib/logstash/compiler/lscl/lscl_grammar.treetop:206-209@6.1`](https://github.com/elastic/logstash/blob/v6.1.3/logstash-core/lib/logstash/compiler/lscl/lscl_grammar.treetop#L206-L209)

But the `ComparisonExpression` is implemented to simply select any child nodes that are one of `Selector`, `Expression`, `ComparisonOperator`, `Number`,  or `String`:

> ~~~ ruby
>   module ComparisonExpression
>     include Helpers
> 
>     def expr
>       lval, comparison_method, rval = self.recursive_select(Selector, Expression, ComparisonOperator, Number, String).map(&:expr)
>       comparison_method.call(source_meta, lval, rval)
>     end
>   end
> ~~~
> -- [`logstash-core/lib/logstash/compiler/lscl.rb:438-445@6.1`](https://github.com/elastic/logstash/blob/v6.1.3/logstash-core/lib/logstash/compiler/lscl.rb#L438-L445)

This leaves out the following `RValue` types, meaning that any expression that includes them will parse, but fail to produce a three-node tuple when selected:
 - `Array`
 - `Hash`
 - `Bareword`
 - `RegRxp`

Additionally, `Expression` is _not_ an `RValue`, and will therefore never be present to be selected.

***

Opening the _grammar_ up to accept `Array` is trivial, but there are some syntax irregularities that make me wary of re-adding full support for Array comparison as part of this bugfix; namely `selector` has a higher precedence than `array`, so while empty arrays (e.g., `[]`) and multi-element arrays (e.g, `["one","two"]` (which aren't valid _selector_s) will parse correctly as arrays, single-element arrays (e.g., `["foo"]` or `[3]`) will be interpreted as a _selector_.

I'm currently working out where the best place is to emit an error for multi-item arrays so that we can present the user with enough context to be useful, and will have a PR ready for that in the next few hours.

**Comment by andrewvc** [comment]
Thanks for the comprehensive summary Ry.

I think this is a good time to bring up the thought that a more regular
Logstash language would be nice. It might be fun to experiment with ideas
for alternate languages for LS configs that are more regular.

I do think we should figure something out for the current language, but I
don't really know if we have any answers, as you've pointed out Ry, we're
painted into a bit of a corner.
